### PR TITLE
[core] Make periodical_runner thread safe.

### PR DESCRIPTION
**Pull request by rynewang**
This PR makes periodical_runner to be thread safe by changing its boolean flag to be atomic. Also makes RedisContext::RunArgvAsync thread safe, and the  RunArgvSync counterpart checking single threadness from the caller. With all these changes we can declare InternalKVManager to be switchable between default io context or a dedicated one.

### CODE REVIEW

File: src/ray/common/asio/periodical_runner.h [hunk 1]
```diff
@@ -49,14 +49,16 @@ class PeriodicalRunner {
...
-                                       const std::string name)
+                                       const std::string &name)
       ABSL_LOCKS_EXCLUDED(mutex_);
 
   instrumented_io_context &io_service_;
   mutable absl::Mutex mutex_;
   std::vector<std::shared_ptr<boost::asio::deadline_timer>> timers_
       ABSL_GUARDED_BY(mutex_);
-  std::shared_ptr<bool> stopped_;
+  // `stopped_` is copied to the timer callback, and may outlive `this`.
```
[the remark this answers is not in the document]
Comment (rynewang) [reply src/ray/common/asio/periodical_runner.h]: done

File: src/ray/common/asio/periodical_runner.cc [hunk 2]
```diff
@@ -74,28 +74,27 @@ void PeriodicalRunner::DoRunFnPeriodically(
...
+      // gcs_server)
+      return;
+    }
+    RAY_CHECK(!error) << error.message();
+    DoRunFnPeriodically(fn, period, timer);
+  });
 }
 
 void PeriodicalRunner::DoRunFnPeriodicallyInstrumented(
     const std::function<void()> &fn,
```
[the remark this answers is not in the document]
Comment (rynewang) [reply src/ray/common/asio/periodical_runner.cc:88]: done

File: src/ray/common/asio/periodical_runner.cc [hunk 3]
```diff
@@ -133,7 +126,7 @@ void PeriodicalRunner::DoRunFnPeriodicallyInstrumented(
           RAY_CHECK(!error) << error.message();
           DoRunFnPeriodicallyInstrumented(fn, period, timer, name);
         },
-        std::move(stats_handle));
+        stats_handle);
```
[the remark this answers is not in the document]
Comment (rynewang) [reply src/ray/common/asio/periodical_runner.cc]: fixed

**Comment by rynewang** [comment]
ready to reveiw @jjyao

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 4]
```diff
@@ -569,15 +581,16 @@ void GcsServer::InitUsageStatsClient() {
 void GcsServer::InitKVManager() {
   // TODO (yic): Use a factory with configs
   std::unique_ptr<InternalKVInterface> instance;
+  auto &io_context = io_context_provider_.GetIOContext<GcsInternalKVManager>();
   switch (storage_type_) {
   case (StorageType::REDIS_PERSIST):
     instance = std::make_unique<StoreClientInternalKV>(
-        std::make_unique<RedisStoreClient>(GetOrConnectRedis()));
+        std::make_unique<RedisStoreClient>(CreateRedisClient(io_context)));
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_server.cc:588]: Should this run in the default io context?

File: src/ray/gcs/gcs_server/gcs_server_io_context_policy.h [hunk 5]
```diff
@@ -40,10 +41,13 @@ struct GcsServerIOContextPolicy {
       return IndexOf("pubsub_io_context");
     } else if constexpr (std::is_same_v<T, syncer::RaySyncer>) {
       return IndexOf("ray_syncer_io_context");
+    } else if constexpr (std::is_same_v<T, GcsInternalKVManager>) {
+      // default io context
+      return -1;
```
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_server_io_context_policy.h:46]: I think this change is out of scope of this PR

File: src/ray/common/asio/asio_util.h [hunk 6]
```diff
@@ -145,7 +146,11 @@ class IOContextProvider {
   instrumented_io_context &GetIOContext() const {
     constexpr int index = Policy::template GetDedicatedIOContextIndex<T>();
     static_assert(
-        index >= -1 && index < Policy::kAllDedicatedIOContextNames.size(),
+        (index == -1) ||
+            (index >= 0 &&
+             static_cast<size_t>(index) < Policy::kAllDedicatedIOContextNames.size()) ||
+            // To show index in compile error...
```
[the remark this answers is not in the document]
Comment (dentiny) [reply src/ray/common/asio/asio_util.h:152]: If precondition is not met, `ray::AlwaysFalseValue<index>` will print out the index value.
Example: https://onlinegdb.com/djkpTXxHl

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 4 above)
Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_server.cc:588]: I think this should go with internal kv's context?

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 7]
```diff
@@ -833,20 +846,12 @@ std::string GcsServer::GetDebugState() const {
...
-    RAY_CHECK(status.ok()) << "Failed to init redis gcs client as " << status;
-
-    // Init redis failure detector.
-    gcs_redis_failure_detector_ = std::make_unique<GcsRedisFailureDetector>(
-        io_context_provider_.GetDefaultIOContext(), redis_client_, []() {
-          RAY_LOG(FATAL) << "Redis connection failed. Shutdown GCS.";
-        });
-    gcs_redis_failure_detector_->Start();
-  }
-  return redis_client_;
```
[the remark this answers is not in the document]
Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_server.cc:849]: Because now we will have 2 redis clients: 1 for internal kv, 1 for others, and they use different ioctx so should not be cached

File: src/ray/common/asio/periodical_runner.cc [hunk 8]
```diff
@@ -38,102 +37,102 @@ void PeriodicalRunner::Clear() {
...
   fn();
   absl::MutexLock lock(&mutex_);
   timer->expires_from_now(period);
   // NOTE: We add the timer period to the enqueue time in order only measure the time in
   // which the handler was elgible to execute on the event loop but was queued by the
   // event loop.
   auto stats_handle = io_service_.stats().RecordStart(name, period.total_nanoseconds());
-  timer->async_wait([this,
+  auto weak_self = weak_from_this();
+  timer->async_wait([weak_self,
```
Comment (dentiny) [inline src/ray/common/asio/periodical_runner.cc:108]: Same here, looks like `weak_self` could be moved; what about `weak_ref = weak_from_this()` in the capture list?

File: src/ray/common/asio/asio_util.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (rynewang) [reply src/ray/common/asio/asio_util.h:152]: yeah, if we got an out of bound index, we want the index number show up in the error message
